[rqt_image_view]Added context menu to hide the upper menu bar

martonmiklos opened the pull request:
This patch adds a simple right mouse button hide menu option to the rqt_image_view:

![kep](https://cloud.githubusercontent.com/assets/1609182/22866237/3fb8a43e-f173-11e6-90f5-54fa93ff0947.png)

If this option is unchecked (by default) the whole toolbar is visible:
![kep](https://cloud.githubusercontent.com/assets/1609182/22866245/65ab1ad2-f173-11e6-9d94-ee3958e53815.png)

When checked the toolbar is hidden:
![kep](https://cloud.githubusercontent.com/assets/1609182/22866252/78abca96-f173-11e6-991e-4e6f2d594e14.png)

...leaving more space on the rqt's ui for another widgets.

The toolbar's visibility state is saved to the settings.

This PR should be the beginning of a PR series what would the following features to rqt:
- Overlaying an ARGB image to the image view
- Display image latency
- Image rotation
- Pausable streaming

The patches above will add significant amount controls to the menu which is better to be hidden.

I have also moved the button icons to the ui file (from C++ code), because it makes more straigthforward to edit them with Qt designer.

dirk-thomas commented:
Thank you for working on this.

Please avoid changing to many things in a single PR. It makes it unnecessary complex to review and see what was changed. I few things from this patch which should be split into separate PRs:

* add context menu to toogle visibility of the menubar
* move the setting of icons from the sources to the ui file (have you made sure that this works in Qt5 as well as Qt 4?)

You should avoid to make any unrelated whitespace changes as well as reordering the elements of the menubar. If you think that the later should be done please keep it separate from other functional changes.

New methods should follow the naming style of the code in the file, e.g. camel-case rather then lowercase-underscore.

martonmiklos commented:
Hi Thomas,

Thank you for the review!

I will separate the icon changes if you want. It is a minor change, basically removing those lines, and setting it in the designer here:
![kep](https://cloud.githubusercontent.com/assets/1609182/23376443/5e7684ee-fd2c-11e6-9bd0-cea710291a28.png)
Yes it is compatible with both Qt4 and Qt5, uic generates essentially the same code just to the setupUI function. 

> You should avoid to make any unrelated whitespace changes as well as reordering the elements of the menubar. If you think that the later should be done please keep it separate from other functional changes.

I have changed the ui file with the designer, the whitespace add/removal is the result of that. I have had to touch that to add the controls to a widget to be able to show/hide all with a single line. 
I will double check the original layout and modify to keep the old organization.

> New methods should follow the naming style of the code in the file, e.g. camel-case rather then lowercase-underscore.

I will refactor the set_controls_visiblity, but I am unsure about the variables. We have currently:
`  QString arg_topic_name;
  ros::Publisher pub_mouse_left_;

  bool pub_topic_custom_;`
What is the meaning of the trailing underscore?

dirk-thomas commented:
> What is the meaning of the trailing underscore?

A trailing underscore indicates a member variable of a class. It is used to clearly separate them from local variables.

martonmiklos commented:
Got it!

- I have changed the capitalization to camel cased, and changed the variable name to have a suffix. 
- Backorganized  the ui file (before - after):
![kep](https://cloud.githubusercontent.com/assets/1609182/23378110/88cd5622-fd32-11e6-8904-6d08d69022b1.png)
Selection on the after image shows the hide/show widget. 
- As you can see the icons are appear on the ui file now. Do you still want to separate that change?

DorianScholz commented:
Having the icon files in the UI file rather than in the code is the nicer solution. But if I remember correctly, it did not work with pyside at the time of writing this code. This might have changed, so please check what the current state is. Thanks for your work on this!

martonmiklos commented:
The rqt_image_view does not uses pyside, but uses C++. 

The ui file is parsed by the uic and the following code is generated:

>         refresh_topics_push_button = new QToolButton(controlsWidget);
        refresh_topics_push_button->setObjectName(QString::fromUtf8("refresh_topics_push_button"));
        QIcon icon1;
        QString iconThemeName = QString::fromUtf8("view-refresh");
        if (QIcon::hasThemeIcon(iconThemeName)) {
            icon1 = QIcon::fromTheme(iconThemeName);
        } else {
            icon1.addFile(QString::fromUtf8("."), QSize(), QIcon::Normal, QIcon::Off);
        }
        refresh_topics_push_button->setIcon(icon1);

DorianScholz commented:
Doh, you're right. Sorry about that, I was thinking of other plugins...

lucasw commented:
If a .perspective is saved containing rqt_image_view windows with hidden menus, does the hidden state get saved into the perspective?

lucasw commented:
Just tried out the branch and it does save the hidden setting correctly.

martonmiklos commented:
The menu hide settings is saved in the same manner as the rest of the settings. I am unsure what do you mean under saving to .perspective. I have created a new perspective from the Perspectives menu and that saves the new setting. What environment do you use? Which version of ROS, Qt and rqt did you tried?

lucasw commented:
It works.  (on Ubuntu 14.04 with Jade and qt4)

martonmiklos commented:
Okay, then in what environment did not it worked?

martonmiklos commented:
Ah I will never comment before my morning coffee again. Somehow I have misread your previous comment as it does not works. :)

martonmiklos commented:
@dirk-thomas

Do you have any additional objections?

dirk-thomas commented:
The merge of another PR resulted in conflicts. Therefore I resolved those and merged your changes in f73be7ffdde27065d12897940ee4de44f5514db1. Please also see d6194290bbdc7c48a241bb15d624cbcbfefc8422 where I changed some of the names.

Thank you for your contribution!

martonmiklos commented:
Thank you so much Dirk!
I will submit some other new features soon!

dirk-thomas commented:
@martonmiklos That is great to hear. You might want to to wait a day or two since I am about to split this monolithic repo into multiple repositories (see https://groups.google.com/d/msg/ros-sig-rqt/CMGkwUndJMI/5AVSksoXBgAJ).

martonmiklos commented:
Thanks for the heads up Dirk! I will wait for the split and I will publish my changes afterwards.
